docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md

```diff
@@ -0,0 +1,271 @@
+
+#### End to End (Prepare & Observe)
+The forecast framework supports specifying metric source and algorithm as needed. So we have a `Prepare` stage to do the
+chores works at the first beginning, converting the interface into forecast framework, creating clients of metric and
+`target` and loading checkpoint to warm up models.
+
+After all stages done in each round, the `Observe` stage shows the result according to the format defined by models.
+For example, a `Statistical Distribution` type of model will show the result as:
+```go
+type DistributionValues struct {
+	Mean              resource.Quantity                `json:"mean,omitempty"`
+	Quantiles         map[string]resource.Quantity     `json:"quantiles,omitempty"`
+	StdDev            resource.Quantity                `json:"stddev,omitempty"`
+	TotalSamplesCount int                              `json:"totalSamplesCount,omitempty"`
+}
+```
+
+### How to assemble forecasting models to support multiple scenarios
+All forecasting stages designed as tool packages, which can be extended with various implementation (e.g. Prometheus 
+or metric-server type `Collector`). We assemble all stages into a `Closure` for each target according to the spec, 
+then a module called `Forecast Runner` will run these `Closure` parallel.
+    
+We define a `Closure` for each target, include the metric and prediction model to use. `Closure` can be run independently
+by `Forecast Runner`.
+
+![forecast-closure](../images/forecasting/forecast-closure.svg)
+
+The top-level are controllers for different scenarios such as resource recommend and interference detection. We define
+CRD for each controller and then the CR will be convert to a `Closure` we defined in forecast framework. The 
+`Forcast Runner` executes all `Closures` and produces the result through observer, which will be convert and write back
+to CRD status field.
+
+![forecasting-controller](../images/forecasting/forecasting-controller.svg)
+
+## Examples
+
+### Detect Outliers for Container CPI Metric
+Container [CPI metric](https://static.googleusercontent.com/media/research.google.com/zh-CN//pubs/archive/40737.pdf) shows
+the runtime quality of applications, which keeps relative stable over serving time. We build a prediction model based on
+historical metric data to determine a container disrupted by others if metric has outliers at present. 
+
+### Profile Resource Usage for FaaS Tasks
+A FaaS service is consist of functions, which receive requests and invoke the executor. Resource usage metrics are exposed
+for `Prometheus` as exporter.
+```
+function_cpu_usage{service="service-word-count", function="f-map-name", slice="xxx"} 1.2
+function_memory_usage{service="service-word-count", function="f-map-name", slice="xxx"} 1024000
+``` 
+We use the forecasting framework to aggregate the metric of each function, and predict the resource usage for right-sizing
+and load aware scheduling.
+
```

Comment:
This is a big topic, please add a section here describing our Milestone (aka our implementation path)